contracts/truefi2/CrvBaseRateOracle.sol

```diff
@@ -0,0 +1,158 @@
+        uint256 idx = histBuffer.insertIndex > 0 ? uint256(histBuffer.insertIndex).sub(1) : bufferSize() - 1;
+        uint256 lastWritten = histBuffer.timestamps[idx];
+        require(block.timestamp >= lastWritten.add(cooldownTime), "CrvBaseRateOracle: Buffer on cooldown");
+        _;
+    }
+
+    constructor(ICurve _curve, uint256 _cooldownTime) public {
+        curve = _curve;
+        cooldownTime = _cooldownTime;
+
+        // fill one field up of the historical buffer
+        // so the first calculateAverageRate call won't return 0
+        histBuffer.baseRates[0] = curve.get_virtual_price();
+        histBuffer.timestamps[0] = block.timestamp;
+        histBuffer.insertIndex++;
+    }
+
+    function bufferSize() public virtual pure returns (uint16) {
+        return 365;
+    }
+
+    /**
+     * @dev Helper function to get contents of the historical buffer
+     */
+    function getHistBuffer()
+        public
+        view
+        returns (
+            uint256[MAX_BUFFER_SIZE] memory,
+            uint256[MAX_BUFFER_SIZE] memory,
+            uint16
+        )
+    {
+        return (histBuffer.baseRates, histBuffer.timestamps, histBuffer.insertIndex);
+    }
+
+    /**
+     * @dev Update the historical buffer:
+     * overwrites the oldest value with current one
+     * and updates its timestamp
+     */
+    function updateRate() public offCooldown {
+        uint256 iidx = histBuffer.insertIndex;
+        histBuffer.timestamps[iidx] = block.timestamp;
+        histBuffer.baseRates[iidx] = curve.get_virtual_price();
+        histBuffer.insertIndex = uint16(iidx.add(1) % bufferSize());
+    }
+
+    /**
+     * @dev Average rate is calculated by taking
+     * the time-weighted average of the curve virtual prices.
+     * Formula is given below:
+     *
+     *           (v + v_{n-1}) / 2 * (t - t_{n-1}) + sum_{i=1}^{n - 1} (v_i + v_{i-1}) / 2 * (t_i - t_{i-1})
+     * avgRate = -------------------------------------------------------------------------------------------
+     *                                                  (t - t_0)
+     *
+     * where v_i, t_i are values of the prices and their respective timestamps
+     * stored in the historical buffer, v is a value of current price ant t is its timestamp.
+     * Index n-1 corresponds to the most recent values and index 0 to the oldest ones.
+     * Notice that whether we are going to use the whole buffer or not
+     * depends on value of timeToCover parameter.
+     * @param timeToCover For how much time average should be calculated
+     * @return Average rate in basis points
+     */
+    function calculateAverageRate(uint256 timeToCover) public view returns (uint256) {
+        require(1 days <= timeToCover && timeToCover <= 365 days, "CrvBaseRateOracle: Expected amount of time in range 1 to 365 days");
+        // estimate how much buffer we need to use
+        uint256 bufferSizeNeeded = timeToCover.div(cooldownTime);
```

Comment:
In practice, I would guess that `updateRate` will be called noticeably later each day, which means that the actual buffer size needed may end up smaller than this value. Perhaps remove this variable entirely, and instead break out of the loop either right before or right after the point when `totalTime > timeToCover`?